File: python/6_19_2023/binary_search.py
# Online Python compiler (interpreter) to run Python online.
# Write Python 3 code in this online editor and run it.
print("Hello world")

import random 
from math import ceil 

def unit_test():
    arr = random.sample(range(0, 1000), 10)
    print(arr)
    arr.sort()
    print(arr)
    for i, key in enumerate(arr):
        found_i = bin_search(arr, key, True)
        print(found_i, i)
        assert found_i == i 

def bin_search(arr, key, is_sorted=False):
    #Verify the array is sorted 
    if not is_sorted:
        arr.sort()
    left = 0
    right = len(arr)
    while left < right:
        mid = (right + left) // 2
        print(left, right, mid, arr[mid])
        if arr[mid] == key:
            return mid 
        elif arr[mid] < key:
            left = mid + 1
        elif arr[mid] > key:
            right = mid - 1
        else:
            print("not possible")
    return -1 
    
unit_test()

File: java/6_19_2023/BinarySearch.java
import java.util.Random;
import java.util.Arrays;

class HelloWorld {
    public static void main(String[] args) {
        System.out.println("Hello, World!");
        UnitTest ut = new UnitTest();
        ut.unitTest();
    }
}

class UnitTest 
{
    public void unitTest()
    {
        int arr[] = new Random().ints(10, 0, 1000).toArray();
        Arrays.sort(arr);
        for(int i=0;i<arr.length;i++)
        {
            System.out.print(arr[i] + " ");
        }
        BinarySearch bs = new BinarySearch();
        for (int i=0; i<arr.length; i++)
        {
            int foundI = bs.search(arr, arr[i], true);
            if (foundI == i)
            {
                //System.out.println("Found it!");
            }
            else
            {
                // If the array contains duplicates only the first key will be found 
                System.out.println("Didn't find it!");
                System.out.println(foundI + " " + i);
            }
                
        }
    }
}

class BinarySearch
{
    public int search(int[] arr, int key, boolean isSorted)
    {
        if (!isSorted)
            Arrays.sort(arr);
        int low = 0;
        int high = arr.length;
        while (low <= high)
        {
            int mid = (low + high) / 2;
            if (arr[mid] == key)
                return mid;
            else if(arr[mid] < key)
                low = mid + 1;
            else if(arr[mid] > key)
                high = mid - 1;
            else
                System.out.println("Not possible");
        }
        return -1; 
    }
}
